### ISSUE TITLE
Try out v2 Preview

### ISSUE THREAD

**Issue by konradpabjan** [actions/download-artifact#23]
The v2-preview of download-artifact is out and we need your help!

You can try it out by using `actions/download-artifact@v2-preview`
Any associated code and documentation can be found here: https://github.com/actions/download-artifact/tree/v2-preview

This issue is for general feedback and to report an bugs/issues during the preview

There is also a `v2-preview` of `upload-artifact`, see: https://github.com/actions/upload-artifact/issues/62

**Warning: At any time during the preview, there may be unannounced changes that can cause things to break. It is recommended not to use the preview of this action in critical workflows**

----

### Historical Context around v2-preview and v1

The `v1` versions of `download-artifact` (and `upload-artifact` ) are plugins that are executed by the runner. The code for `v1` can be found here: https://github.com/actions/runner/tree/master/src/Runner.Plugins/Artifact

The `v1` code is written in C# and is tightly coupled to the runner, it also uses special APIs that only the runner can use to interact with artifacts. If any changes or updates had to be made related to artifacts, they had to done on the runner and a new release had to roll out that would take a significant amount of time. With v2, there is no dependency on the runner so it will be much easier and faster to make changes and accept community contributions (right now it was pretty much impossible).

The `v2-preview` of download-artifact has been rewritten from scratch using Typescript with a new set of APIs that allow it interact with artifacts (previously only the runner could do this). There is a new NPM package called `@actions/artifact` that contains the core functionality to interact with artifacts (which this action uses for the most part). This NPM package is hosted in the `actions/toolkit` repo so anyone can use this to interact with artifacts when developing actions. You can find the package here (lots of documentation and extra info):
https://www.npmjs.com/package/@actions/artifact
https://github.com/actions/toolkit/tree/master/packages/artifact

### Further work

Since `v2-preview` is effectively a total rewrite from `v1`, There is a huge potential for bugs so it needs to be tested thoroughly before creating an actual `v2` release. We need help testing the core functionality which includes:
- Downloading a single artifact
- Downloading multiple artifacts at once

There will be no new features added as part of the `v2-preview`, we need to test the core functionality first and once a `v2` release is out, then we can start chipping away at issues/features that we have been unable to address with `v1`🙂 For the moment, please don't submit PRs for any new features to the `v2-preview` branch.

**Comment by spookyGh0st** [comment]
getting error `##[error]read ECONNRESET` when downloading a single artifact

my download  action

```yaml
      - name: get sirup.jar
        uses: actions/download-artifact@v2-preview
        with:
          name: sirup.jar
```

v1 works fine, only getting this error when using v2-preview.

**Comment by lovasoa** [comment]
Does the v2 solve https://github.com/actions/download-artifact/issues/11 ? It would be really nice to have the downloaded file path as the output of the action.

**Comment by konradpabjan** [comment]
I left a comment in the issue @lovasoa We can easily add the feature, though what do we do with the download file path if there are multiple artifacts being downloaded?

**Comment by konradpabjan** [comment]
I'm seeing some issues with the `v2-preview` (upload is going great, but download is having some problems). Working on an update

**Comment by CKGrafico** [comment]
Not sure if reported but in some point when I download an uploaded artifact all the txt files were converted to BINARY files, I solved using upload v2 but download v1, really strange.

```
      - name: Get the artifacts
        uses: actions/download-artifact@v2-preview
        with:
          name: ${{env.ARTIFACT_NAME}}
          path: ./dotnetcorewebapp

```
now:
```
      - name: Get the artifacts
        uses: actions/download-artifact@v1
        with:
          name: ${{env.ARTIFACT_NAME}}
          path: ./dotnetcorewebapp
```

**Comment by konradpabjan** [comment]
Thanks for the heads up @CKGrafico 

I suspect it's a problem with gzip decompression which is used for large files during upload. Taking a look.

**Comment by CaselIT** [comment]
I think I'm having the same problem with the compression.

What I want to do is:
job1:
- create a python wheel in a matrix
- upload the single wheels

job2:
- download all artifact wheels
- upload a single artifact with all of them.

The job2 downloads the wheels but then changes them. I'm not sure what it happens, but it seems that instead of unzipping the content of the artifact it renames it with the name of the file inside.

Here is a run of the job: https://github.com/CaselIT/sqlalchemy/actions/runs/65246582
The two artifact created by job1 are  `single-wheel-ubuntu-latest-3.7` and `single-wheel-ubuntu-latest-2.7`.
They both contain a single file with the created wheel.
The `all-wheels` artifacts contains the two wheels, but the two files are different from the files in the separated artifacts.

**Comment by zhangyoufu** [comment]
I have a strategy.matrix that uploads a lot of artifacts. I wish there is an input parameter for download-artifact that can merge all artifacts into a single directory. Currently I achieve this by hard-linking using `cp -l`.

```
    - name: Download all artifacts
      uses: actions/download-artifact@v2-preview
    - name: Merge all artifacts
      run: mkdir repo; cp -lr repo-*/* repo
```

**Comment by zhangyoufu** [comment]
I have some gzip-compressed files inside artifact, got gzip-compressed again during download.

**Comment by yarikoptic** [comment]
confirming: artifacts are gzip-compressed (see e.g. log with `file` being ran on all of them: https://github.com/datalad/datalad-extensions/runs/548784573#step:3:19), although when I just download artifacts and extract the .zip archive -- they are all good.  If I explicitly extract using gzip - content seems to be ok.
My wild guess -- may be the http transport while transferring content to the workflow uses gzip ContentEncoding, but then receiving end doesn't extract?

**Comment by konradpabjan** [comment]
I've pinpointed the issue and got everything working. An update needs to roll out for the `@actions/http-client` package and then the also to the `@actions/artifact` package so it will take a little time for the changes to show up here. For those curious there is some pretty weird stuff going on with content encodings. Going from `gzip` to `utf-16` back to `gzip` results in some permanent information being lost along the way. Took some time to pinpoint this.

https://github.com/actions/http-client/pull/15
https://github.com/actions/toolkit/pull/396

**Comment by dawnbass68** [comment]
I am probably over my head with all you very smart people . Im sorry but I
really do not knoow what you are asking me to do.  I am more than happy to
contribute want to in fact . I do appreciate you and hope I have not caused
you any undo pain.

**Comment by kevinrobinson** [comment]
Thanks @konradpabjan!  I'm seeing the same things others have reported here, where the UI download sends a zip file with the artifiacts, but when downloading the artifacts appear with correct names but are themselves gzipped.

From reading through https://github.com/actions/http-client/pull/15, it sounds like the approach you outlined here is not necessary to do first, and once https://github.com/actions/toolkit/pull/396 lands it will resolve this issue?  Thanks for all your work on this! 👍

**Comment by konradpabjan** [comment]
Version `0.3.0` of the `@actions/artifact` package is out with a whole bunch of fixes (most notably the fixe related to gzip decompression during download). I ended up not having to update anything on the `@actions/http-client` side: https://www.npmjs.com/package/@actions/artifact/v/0.3.0

Full changelog: https://github.com/actions/toolkit/blob/master/packages/artifact/RELEASES.md

I've updated the `v2-preview` branch to use the latest version of this package so it should be good to go now 😀 I've been testing by uploading about 50MB of random `node_modules` and then running a binary comparison against the downloaded artifacts and everything appears to be correct. If there are any problems please let me know.

**Comment by zhangyoufu** [comment]
LGTM. The gzip issue was resolved.

**Comment by konradpabjan** [comment]
Versions `0.3.1` of `@actions/artifact` rolled out with some minor updates: https://www.npmjs.com/package/@actions/artifact/v/0.3.0
I've updated the `v2-preview` branch with the latest package

Other than that everything is looking 👍 If no new concerns come up then next week we are planning on officially releasing v2 (the v2-preview branch will be merged into master and tags will be updated) Expect the v2-preview branch to be deleted shortly after so heads up and this issue will be closed. You can also expect some minor changelog to show up: https://github.blog/changelog/

**Comment by webknjaz** [comment]
@konradpabjan how about supporting `name` with some sort of a wildcard/glob so that we could download artifacts matching certain pattern rather than including unnecessary things too?

**Comment by konradpabjan** [comment]
@webknjaz I'll most definitely be revisiting https://github.com/actions/download-artifact/issues/6 as the current download-all option isn't ideal for all users. 

Once a list API for artifacts is available (see https://github.com/actions/toolkit/issues/379) it should fairly straightforward to implement some sort of wildcard matching so users aren't limited to only a downloading a single artifact or all artifacts

**Comment by konradpabjan** [comment]
`v2` has been merged into master and published! Thank you everyone for the feedback and testing.  https://github.com/marketplace/actions/download-a-build-artifact

You can now directly use it by doing `actions/download-artifact@v2`. 

Heads up, the `@v2-preview` branch will be deleted soon (probably a week or so, I don't want to keep it around for long). If you're using `@v2-preview` in your workflow, please switch over to `@v2` or your workflow will break.

**Comment by cfclrk** [comment]
`actions/download-artifact@v2` (as it currently states in the README) didn't work for me, I had to use:
`actions/download-artifact@v2.0`

**Comment by konradpabjan** [comment]
@cfclrk apologies, I incorrectly set the tag to 2.0 instead of 2. I realized after about an hour, it should be good now
